cpp/src/io/json/json_tree.cu

```diff
@@ -254,12 +321,487 @@ tree_meta_t get_tree_representation(device_span<PdaTokenT const> tokens,
+    auto it = key_map.find(node_id, d_hasher, d_equal);
+    return (it == key_map.end()) ? size_type{0} : it->second.load();
+  };
+  // convert field nodes to node indices, and other nodes to enum value.
+  rmm::device_uvector<size_type> node_type(num_nodes, stream);
+  thrust::tabulate(rmm::exec_policy(stream),
+                   node_type.begin(),
+                   node_type.end(),
+                   [node_categories = d_tree.node_categories.data(),
+                    is_field_name_node,
+                    get_hash_value] __device__(auto node_id) -> size_type {
+                     if (is_field_name_node(node_id))
+                       return static_cast<size_type>(NUM_NODE_CLASSES) + get_hash_value(node_id);
+                     else
+                       return static_cast<size_type>(node_categories[node_id]);
+                   });
+  return node_type;
+}
+
+/**
+ * @brief Translates sorted parent_node_ids to parent_indices with indices from scatter_indices
+ *
+ * @param scatter_indices The sorted order of parent_node_ids
+ * @param parent_node_ids The sorted parent_node_ids
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @return Translated parent_indices pointing to sorted node_ids positions
+ */
+rmm::device_uvector<NodeIndexT> translate_sorted_parent_node_indices(
+  device_span<size_type const> scatter_indices,
+  device_span<NodeIndexT const> parent_node_ids,
+  rmm::cuda_stream_view stream)
+{
+  auto const num_nodes = scatter_indices.size();
+  auto gather_indices  = cudf::detail::scatter_to_gather(
+    scatter_indices.begin(), scatter_indices.end(), num_nodes, stream);
+
+  rmm::device_uvector<NodeIndexT> parent_indices(num_nodes, stream);
+  // gather, except parent sentinels
+  thrust::transform(rmm::exec_policy(stream),
+                    parent_node_ids.begin(),
+                    parent_node_ids.end(),
+                    parent_indices.begin(),
+                    [gather_indices = gather_indices.data()] __device__(auto parent_node_id) {
+                      return (parent_node_id == parent_node_sentinel)
+                               ? parent_node_sentinel
+                               : gather_indices[parent_node_id];
+                    });
+  return parent_indices;
+};
+
+std::pair<rmm::device_uvector<NodeIndexT>, rmm::device_uvector<NodeIndexT>> generate_column_id(
+  device_span<size_type> node_type,        // level sorted
+  device_span<NodeIndexT> parent_indices,  // level sorted
+  device_span<size_type> scatter_indices,  // level sorted
+  host_span<size_type const> level_boundaries,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr)
+{
+  // pre-condition: node_type is level stable sorted. (stable in node_id order)
+  // parent_indices is level stable sorted.
+  // scatter_indices is level stable sorted.
+  // post-condition: returned column_id, parent_col_id are level sorted, scatter_indices holds the
```

Comment:
Yes. modified in place. 
No additional information is returned by `scatter_indices`. It is modified in place and back to original order.
Note that `device_span<size_type>`, not `device_span<size_type const>`. But I understand it's strange.

`scatter_indices` is used at different steps, and it helps re-order across different steps. I can copy it here, but adds extra memory.
will come up with a better way to avoid `scatter_indices` modification.

Reply:
replaced it internally with a sequence as scatter_indices. This will help with scatter in per-level-processing too.